python-threatexchange/threatexchange/fetcher/apis/ncmec_api.py

```diff
@@ -0,0 +1,232 @@
+from threatexchange.signal_type.pdq import PdqSignal
+
+
+_API_NAME: str = "ncmec"
+
+
+@dataclass
+class NCMECCheckpoint(
+    state.FetchCheckpointBase,
+):
+    """
+    NCMEC primarily revolves around polling the timestamp.
+
+    NCMEC IDs seem to stay around forever, so no need for is_stale()
+    """
+
+    max_timestamp: int
+
+    def get_progress_timestamp(self) -> t.Optional[int]:
+        return self.max_timestamp
+
+    @classmethod
+    def from_ncmec_fetch(cls, response: api.GetEntriesResponse) -> "NCMECCheckpoint":
+        return cls(response.max_timestamp)
+
+
+@dataclass
+class _NCMECCollabConfigRequiredFields:
+    environment: api.NCMECEnvironment
+
+
+@dataclass
+class NCMECCollabConfig(
+    CollaborationConfigWithDefaults,
+    _NCMECCollabConfigRequiredFields,
+):
+    api: str = field(init=False, default=_API_NAME)
+
+
+@dataclass
+class NCMECEntryMetadata(state.FetchedSignalMetadata):
+    """
+    Placeholder to store entries from the API
+
+    TODO: It turns out I was not smart enough and build the API interface
+          poorly, which NCMEC has exposed.
+          The chain of Delta => Dict[(type, hash): ?Metadata] => Store => Index
+          breaks because NCME doesn't map easily to a unique (type, hash) pairing.
+          Instead, it has its own unique keys (esp_id, entry_id).
+
+          The fix is to seperate out the concerns needed for UpdateDelta to work
+          (any unique key, a combining function) from the concerns needed for Index
+          to work (group by signal type => Metadata).
+    """
+
+    entry: api.NCMECEntryUpdate
+
+    def get_as_opinions(self) -> t.List[state.SignalOpinion]:
+        raise NotImplementedError(
+            "Placeholder while I figure out the correct model for NCMEC"
+        )
+
+
+@dataclass
+class NCMECSignalMetadata(state.FetchedSignalMetadata):
+    """
+    NCMEC metadata includes who uploaded it, as well as what they tagged.
+
+    The NCMEC API has no concept of false positives - every entry is reported.
+    """
+
+    member_entries: t.Dict[int, t.Set[str]]
+
+    def get_as_opinions(self) -> t.List[state.SignalOpinion]:
+        return [
+            state.SignalOpinion(
+                member_id, state.SignalOpinionCategory.TRUE_POSITIVE, tags
+            )
+            for member_id, tags in self.member_entries.items()
+        ]
+
+
+def _get_conversion(
+    signal_types: t.Sequence[t.Type[SignalType]],
+) -> t.Mapping[t.Tuple[api.NCMECEntryType, str], t.Type[SignalType]]:
+    ret: t.Dict[t.Tuple[api.NCMECEntryType, str], t.Type[SignalType]] = {}
+    if VideoMD5Signal in signal_types:
+        ret[api.NCMECEntryType.video, "md5"] = VideoMD5Signal
+    if PdqSignal in signal_types:
+        ret[api.NCMECEntryType.image, "pdq"] = PdqSignal
+    for st in signal_types:
+        if st.get_name() == "photodna":
+            ret[api.NCMECEntryType.image, "pdna"] = st
+            break
+    return ret
```

Comment:
I need a better way to handle this long term. You end up with a problem where you might have one of the signal type or API. 

This API might be slightly different since it's included in the base threatexchange library, and using string reflection here is okay, but I dunno